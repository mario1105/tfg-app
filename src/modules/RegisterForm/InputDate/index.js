import React, { useEffect } from 'react';
import {
  Input, FormControl, InputLabel, makeStyles,
} from '@material-ui/core';
import PropTypes from 'prop-types';
import Typography from 'components/Typography';

const useStyles = makeStyles(() => ({
  formControlContainer: {
    display: 'flex',
    marginBottom: '.5em'
  },
}));

const setFocus = (inputId) => {
  setTimeout(() => document.getElementById(inputId).focus(), 1);
};

const InputDate = ({
  date, errors, handleFieldsChange, handleFieldsErrors, resetErrors, ...props
}) => {
  const { day, month, year } = date;

  useEffect(
    () => {
      if (day.length === 2) {
        if (month.length !== 2) setFocus('month-input');
        else if (month.length === 2 && year.length !== 4) setFocus('year-input');
      } else if (day.length === 0 && month.length === 2) setFocus('year-input');
    },
    [day, month],
  );

  const classes = useStyles();

  return (
    <div>
      <div className={classes.formControlContainer}>
        <FormControl data-test-id="date-field-day-form-control" error={errors.day} {...props}>
          <InputLabel>DD</InputLabel>
          <Input
            data-test-id="date-field-day-input"
            id={'day-input'}
            type={'number'}
            className={'pureNumberInput'}
            autoComplete="off"
            value={day}
            onChange={handleFieldsChange('day')}
            onBlur={() => handleFieldsErrors('day')}
            onFocus={() => resetErrors('day')}
            onInput={(e) => {
              e.target.value = e.target.value.slice(0, 2);
            }}
          />
        </FormControl>
        <FormControl data-test-id="date-field-month-form-control" error={errors.month} {...props}>
          <InputLabel>MM</InputLabel>
          <Input
            data-test-id="date-field-month-input"
            id={'month-input'}
            type={'number'}
            className={'pureNumberInput'}
            autoComplete="off"
            value={month}
            onChange={handleFieldsChange('month')}
            onBlur={() => handleFieldsErrors('month')}
            onFocus={() => resetErrors('month')}
            onInput={(e) => {
              e.target.value = e.target.value.slice(0, 2);
            }}
          />
        </FormControl>
        <FormControl data-test-id="date-field-year-form-control" error={errors.year} {...props}>
          <InputLabel>YYYY</InputLabel>
          <Input
            data-test-id="date-field-year-input"
            id={'year-input'}
            type={'number'}
            className={'pureNumberInput'}
            autoComplete="off"
            value={year}
            onChange={handleFieldsChange('year')}
            onBlur={() => handleFieldsErrors('year')}
            onFocus={() => resetErrors('year')}
            onInput={(e) => {
              e.target.value = e.target.value.slice(0, 4);
            }}
          />
        </FormControl>
      </div>
      <Typography variant="subtitle2" gutterBottom>
        Date of Birth
      </Typography>
    </div>

  );
};

InputDate.propTypes = {
  date: PropTypes.shape({
    day: PropTypes.string,
    month: PropTypes.string,
    year: PropTypes.string
  }),
  errors: PropTypes.shape({
    day: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    month: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    year: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  }),
  handleFieldsChange: PropTypes.func,
  handleFieldsErrors: PropTypes.func,
  resetErrors: PropTypes.func,
};

export default InputDate;
