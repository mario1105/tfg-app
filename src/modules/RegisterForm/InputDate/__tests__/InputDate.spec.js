import React from 'react';
import { shallow } from 'enzyme';
import InputDate from '../index';

const defaultProps = {
  date: {
    day: '11',
    month: '11',
    year: '1999'
  },
  errors: {
    day: false,
    month: false,
    year: false
  },
  handleFieldsChange: () => null,
  handleFieldsErrors: () => null,
  resetErrors: () => null,
};

const renderComponent = (newProps) => {
  const props = {
    ...defaultProps,
    ...newProps,
  };

  return (shallow(<InputDate {...props} />));
};

let wrapper;
describe('InputDate', () => {
  describe('when there are no errors', () => {
    beforeEach(() => {
      wrapper = renderComponent();
    });

    it('renders as expected', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('sets the input values with the correct prop', () => {
      expect(wrapper.find('[data-test-id~="date-field-day-input"]').prop('value')).toBe(defaultProps.date.day);
      expect(wrapper.find('[data-test-id~="date-field-month-input"]').prop('value')).toBe(defaultProps.date.month);
      expect(wrapper.find('[data-test-id~="date-field-year-input"]').prop('value')).toBe(defaultProps.date.year);
    });
  });
  describe('when there are errors', () => {
    it('renders as expected', () => {
      expect(renderComponent({
        errors: {
          day: true,
          month: true,
          year: true
        },
      })).toMatchSnapshot();
    });
  });
});
