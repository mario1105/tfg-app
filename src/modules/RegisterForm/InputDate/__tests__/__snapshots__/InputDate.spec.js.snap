// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputDate when there are errors renders as expected 1`] = `
<div>
  <div
    className="makeStyles-formControlContainer-1"
  >
    <WithStyles(ForwardRef(FormControl))
      data-test-id="date-field-day-form-control"
      error={true}
    >
      <WithStyles(ForwardRef(InputLabel))>
        DD
      </WithStyles(ForwardRef(InputLabel))>
      <WithStyles(ForwardRef(Input))
        autoComplete="off"
        className="pureNumberInput"
        data-test-id="date-field-day-input"
        id="day-input"
        onBlur={[Function]}
        onChange={null}
        onFocus={[Function]}
        onInput={[Function]}
        type="number"
        value="11"
      />
    </WithStyles(ForwardRef(FormControl))>
    <WithStyles(ForwardRef(FormControl))
      data-test-id="date-field-month-form-control"
      error={true}
    >
      <WithStyles(ForwardRef(InputLabel))>
        MM
      </WithStyles(ForwardRef(InputLabel))>
      <WithStyles(ForwardRef(Input))
        autoComplete="off"
        className="pureNumberInput"
        data-test-id="date-field-month-input"
        id="month-input"
        onBlur={[Function]}
        onChange={null}
        onFocus={[Function]}
        onInput={[Function]}
        type="number"
        value="11"
      />
    </WithStyles(ForwardRef(FormControl))>
    <WithStyles(ForwardRef(FormControl))
      data-test-id="date-field-year-form-control"
      error={true}
    >
      <WithStyles(ForwardRef(InputLabel))>
        YYYY
      </WithStyles(ForwardRef(InputLabel))>
      <WithStyles(ForwardRef(Input))
        autoComplete="off"
        className="pureNumberInput"
        data-test-id="date-field-year-input"
        id="year-input"
        onBlur={[Function]}
        onChange={null}
        onFocus={[Function]}
        onInput={[Function]}
        type="number"
        value="1999"
      />
    </WithStyles(ForwardRef(FormControl))>
  </div>
  <WithStyles(ForwardRef(Typography))
    gutterBottom={true}
    style={
      Object {
        "color": "white",
      }
    }
    variant="subtitle2"
  >
    Date of Birth
  </WithStyles(ForwardRef(Typography))>
</div>
`;

exports[`InputDate when there are no errors renders as expected 1`] = `
<div>
  <div
    className="makeStyles-formControlContainer-1"
  >
    <WithStyles(ForwardRef(FormControl))
      data-test-id="date-field-day-form-control"
      error={false}
    >
      <WithStyles(ForwardRef(InputLabel))>
        DD
      </WithStyles(ForwardRef(InputLabel))>
      <WithStyles(ForwardRef(Input))
        autoComplete="off"
        className="pureNumberInput"
        data-test-id="date-field-day-input"
        id="day-input"
        onBlur={[Function]}
        onChange={null}
        onFocus={[Function]}
        onInput={[Function]}
        type="number"
        value="11"
      />
    </WithStyles(ForwardRef(FormControl))>
    <WithStyles(ForwardRef(FormControl))
      data-test-id="date-field-month-form-control"
      error={false}
    >
      <WithStyles(ForwardRef(InputLabel))>
        MM
      </WithStyles(ForwardRef(InputLabel))>
      <WithStyles(ForwardRef(Input))
        autoComplete="off"
        className="pureNumberInput"
        data-test-id="date-field-month-input"
        id="month-input"
        onBlur={[Function]}
        onChange={null}
        onFocus={[Function]}
        onInput={[Function]}
        type="number"
        value="11"
      />
    </WithStyles(ForwardRef(FormControl))>
    <WithStyles(ForwardRef(FormControl))
      data-test-id="date-field-year-form-control"
      error={false}
    >
      <WithStyles(ForwardRef(InputLabel))>
        YYYY
      </WithStyles(ForwardRef(InputLabel))>
      <WithStyles(ForwardRef(Input))
        autoComplete="off"
        className="pureNumberInput"
        data-test-id="date-field-year-input"
        id="year-input"
        onBlur={[Function]}
        onChange={null}
        onFocus={[Function]}
        onInput={[Function]}
        type="number"
        value="1999"
      />
    </WithStyles(ForwardRef(FormControl))>
  </div>
  <WithStyles(ForwardRef(Typography))
    gutterBottom={true}
    style={
      Object {
        "color": "white",
      }
    }
    variant="subtitle2"
  >
    Date of Birth
  </WithStyles(ForwardRef(Typography))>
</div>
`;
